### PR TITLE
[SPARK-20889][SparkR] Grouped documentation for WINDOW column methods

### PR DESCRIPTION
**Pull request by actuaryzhang**
## What changes were proposed in this pull request?

Grouped documentation for column window methods.

### CODE REVIEW

**Comment by actuaryzhang** [comment]
Ahh, forgot about the window functions. This is actually the last set...  @felixcheung @HyukjinKwon 

image
image
image

**Comment by actuaryzhang** [comment]
jenkins, retest this please

File: R/pkg/R/functions.R [hunk 1]
```diff
@@ -3083,11 +3011,10 @@ setMethod("rank",
             column(jc)
           })
 
-# Expose rank() in the R base package
-#' @param x a numeric, complex, character or logical vector.
-#' @param ... additional argument(s) passed to the method.
-#' @name rank
-#' @rdname rank
+#' @details
+#' \code{rank}: Exposes \code{rank()} in the R base package. In this case, \code{x}
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: I'm actually mixed in whether we should group the doc for this.
I guess our hands are tied and we have to have documentation for it

File: R/pkg/R/functions.R [hunk 2]
```diff
@@ -2844,27 +2869,16 @@ setMethod("ifelse",
...
-#'
-#' Window function: returns the cumulative distribution of values within a window partition,
+#' @details
+#' \code{cume_dist}: Returns the cumulative distribution of values within a window partition,
 #' i.e. the fraction of rows that are below the current row.
-#'
-#'   N = total number of rows in the partition
-#'   cume_dist(x) = number of values before (and including) x / N
-#'
+#' N = total number of rows in the partition
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: these formula esp. with multiple lines becomes very hard to read.
any idea?

File: R/pkg/R/functions.R [hunk 3]
```diff
@@ -2903,34 +2907,16 @@ setMethod("dense_rank",
...
+#' \code{lag}: Returns the value that is \code{offset} rows before the current row, and
 #' \code{defaultValue} if there is less than \code{offset} rows before the current row. For example,
 #' an \code{offset} of one will return the previous row at any given point in the window partition.
-#'
 #' This is equivalent to the \code{LAG} function in SQL.
 #'
-#' @param x the column as a character string or a Column to compute on.
-#' @param offset the number of rows back from the current row from which to obtain a value.
-#'               If not specified, the default is 1.
 #' @param defaultValue (optional) default to use when the offset row does not exist.
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: perhaps move this to common block above?

File: R/pkg/R/generics.R [hunk 4]
```diff
@@ -1013,9 +1013,9 @@ setGeneric("create_map", function(x, ...) { standardGeneric("create_map") })
 #' @name NULL
 setGeneric("hash", function(x, ...) { standardGeneric("hash") })
 
-#' @param x empty. Should be used with no argument.
```
Comment (felixcheung) [inline R/pkg/R/generics.R]: document these ones that should be use without argument
there are 4 of these

File: R/pkg/R/functions.R [hunk 5]
```diff
@@ -200,6 +200,31 @@ NULL
...
+#' @param ... additional argument(s).
+#' @name column_window_functions
+#' @rdname column_window_functions
+#' @family window functions
+#' @examples
+#' \dontrun{
+#' # Dataframe used throughout this doc
+#' df <- createDataFrame(cbind(model = rownames(mtcars), mtcars))
+#' ws <- orderBy(windowPartitionBy("am"), "hp")
+#' tmp <- mutate(df, dist = over(cume_dist(), ws), dense_rank = over(dense_rank(), ws),
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: is this comment still useful, or obvious without it?
```
#'   # Get ntile group id (1-4) for hp
```

**Review by felixcheung — CHANGES_REQUESTED** [review]

File: R/pkg/R/functions.R (hunk 5 above)
Comment (actuaryzhang) [inline R/pkg/R/functions.R]: OK. added back

File: R/pkg/R/functions.R (hunk 2 above)
Comment (actuaryzhang) [inline R/pkg/R/functions.R]: Fixed with better doc.

File: R/pkg/R/functions.R (hunk 3 above)
Comment (actuaryzhang) [inline R/pkg/R/functions.R]: Done.

File: R/pkg/R/functions.R (hunk 1 above)
Comment (actuaryzhang) [inline R/pkg/R/functions.R]: Ye, actually we don need to doc this. We only need to add an alias and that should satisfy the R requirement. User can still use the base doc.

File: R/pkg/R/generics.R (hunk 4 above)
Comment (actuaryzhang) [inline R/pkg/R/generics.R]: added.

File: R/pkg/R/functions.R [hunk 6]
```diff
@@ -2871,10 +2874,10 @@ setMethod("ifelse",
 
 #' @details
 #' \code{cume_dist}: Returns the cumulative distribution of values within a window partition,
-#' i.e. the fraction of rows that are below the current row.
-#' N = total number of rows in the partition
-#' cume_dist(x) = number of values before (and including) x / N
+#' i.e. the fraction of rows that are below the current row:
+#' number of values before (and including) x / total number of rows in the partition.
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: with that many words it seems the fact this is a formula is lost. how about this like the other formula
```

(number of values before and including x) / (total number of rows in the partition)
```

File: R/pkg/R/functions.R [hunk 7]
```diff
@@ -2844,27 +2872,16 @@ setMethod("ifelse",
...
-#'
-#'   N = total number of rows in the partition
-#'   cume_dist(x) = number of values before (and including) x / N
-#'
+#' @details
+#' \code{cume_dist}: Returns the cumulative distribution of values within a window partition,
+#' i.e. the fraction of rows that are below the current row:
+#' number of values before (and including) x / total number of rows in the partition.
 #' This is equivalent to the \code{CUME_DIST} function in SQL.
+#' This should be used with no argument.
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: in the earlier PR, you have `The method should be used with no argument.`

**Comment by actuaryzhang** [comment]
OK, docs are now updated as you suggested.

**Comment by actuaryzhang** [comment]
jenkins, retest this please

File: R/pkg/R/functions.R [hunk 8]
```diff
@@ -2875,9 +2875,9 @@ setMethod("ifelse",
 #' @details
 #' \code{cume_dist}: Returns the cumulative distribution of values within a window partition,
 #' i.e. the fraction of rows that are below the current row:
-#' number of values before (and including) x / total number of rows in the partition.
+#' (number of values before and including x) / (total number of rows in the partition)
 #' This is equivalent to the \code{CUME_DIST} function in SQL.
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: very minor nit - without some kind of separator, I think roxygen will try to put the 2 lines together, like

```
w the current row: (number of values before and including x) / (total number of rows in the partition) This is equivalent to the \code{CUME_DIST} function in SQL.
```

in general it's hard to do formula in doc. could you check?

Comment (actuaryzhang) [inline R/pkg/R/functions.R]: This is not a formula, right? Thought this is pretty clear. Not sure what is the ask. I can just add a period after the `partition)`.

Comment (felixcheung) [inline R/pkg/R/functions.R]: sure, that should be fine - I mean that it is a equation of some sort something divided by something. I guess that is hard to read in R doc.

**Review by felixcheung — APPROVED** [review]
LGTM

**Comment by felixcheung** [comment]
merged to master

**Pull request closed without merging**